src/cca/scripts/uproxy_server.ts

```diff
@@ -56,25 +36,146 @@ export class UproxyServer implements Server {
   }
 }
 
+// https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API
+function getIsStorageAvailable() {
+  try {
+    const storage = window['localStorage'];
+    const x = '__storage_test__';
+    storage.setItem(x, x);
+    storage.removeItem(x);
+    return true;
+  } catch(e) {
+    return false;
+  }
+}
+const isStorageAvailable = getIsStorageAvailable();
+
+// Name by which servers are saved to storage.
+const CONTACTS_STORAGE_KEY = 'cloud-server-invites';
+
 export class UproxyServerRepository implements ServerRepository {
+  // core must already be logged into social networks.
   constructor(private core: CoreConnector, private vpnDevice: VpnDevice) {
-    this.core.login({
-      network: 'Cloud',
-      loginType: uproxy_core_api.LoginType.INITIAL,
+    if (!isStorageAvailable) {
+      console.warn('localStorage unavailable - contacts will not be saved');
+      return;
+    }
+  }
+
+  public onServerCallback :OnServerCallback;
+  public onServer(callback:OnServerCallback) {
+    this.onServerCallback = callback;
+    return this;
+  }
+
+  public restore() {
```

Comment:
Since this restore() function only makes sense at construction, it may make sense to remove this method from the class, and place it in a factory function instead.